# Easy

In [14]:
import pandas as pd

train = pd.read_pickle('../data/train.pickle')
dataframe_train = pd.DataFrame(train['features'].reshape(train['features'].shape[0], -1))
dataframe_train.head()

0     1     2     3     4     5     6     7     8     9     ...  3062  \
0    28    25    24    27    24    23    27    24    22    27  ...    22   
1    28    24    24    26    23    23    27    24    24    28  ...    22   
2    29    25    25    29    26    26    30    27    27    30  ...    22   
3    31    28    26    29    26    24    30    27    26    29  ...    22   
4    33    27    25    31    25    24    32    26    25    31  ...    21   

   3063  3064  3065  3066  3067  3068  3069  3070  3071  
0    27    24    22    27    23    21    26    22    20  
1    27    23    22    27    24    23    29    26    25  
2    29    25    23    28    24    22    30    26    24  
3    28    23    22    27    23    22    27    23    21  
4    25    22    20    26    22    19    27    22    19  

[5 rows x 3072 columns]

In [16]:
test = pd.read_pickle('../data/test.pickle')
dataframe_test = pd.DataFrame(test['features'].reshape(test['features'].shape[0], -1))
dataframe_test.head()

0     1     2     3     4     5     6     7     8     9     ...  3062  \
0   116   139   174   116   137   171   118   138   172   117  ...   164   
1    59    70    61    87    80    63    92    81    63    96  ...    58   
2    52    40    38    51    39    37    50    38    37    50  ...    34   
3   132   125    87   100    92    60    82    73    34    96  ...   152   
4    37    34    30    42    38    32    46    40    36    44  ...    35   

   3063  3064  3065  3066  3067  3068  3069  3070  3071  
0   115   138   167   117   138   167   114   140   170  
1    62    59    62    73    67    70    75    70    70  
2    41    40    38    44    43    43    47    46    44  
3   235   239   196   255   251   217   255   240   180  
4    46    41    38    48    43    40    49    45    42  

[5 rows x 3072 columns]

In [20]:
from sklearn.linear_model import LogisticRegression

model = LogisticRegression(max_iter=500)
model.fit(dataframe_train, train['labels'])

/Users/elizabeth/anaconda3/lib/python3.11/site-packages/sklearn/linear_model/_logistic.py:460: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  n_iter_i = _check_optimize_result(


LogisticRegression(max_iter=500)

In [21]:
from sklearn.metrics import accuracy_score, precision_score, recall_score

y_pred = model.predict(dataframe_test)

# Оценка качества модели с использованием метрик по данным из test.pickle
accuracy = accuracy_score(y_pred, test['labels'])
precision = precision_score(y_pred, test['labels'], average='macro')
recall = recall_score(y_pred, test['labels'], average='macro')

print(f"Accuracy: {accuracy}")
print(f"Precision: {precision}")
print(f"Recall: {recall}")

Accuracy: 0.8434679334916865
Precision: 0.777472725304879
Recall: 0.810423128479271


In [22]:
from sklearn.metrics import confusion_matrix

conf_matrix = confusion_matrix(y_pred, test['labels'])
conf_matrix.shape

(43, 43)

In [34]:
conf_matrix = confusion_matrix(y_pred, test['labels'])
errors = []  # сделали список с ошибками

for i in range(len(conf_matrix)):
    for j in range(i + 1, len(conf_matrix)):
        errors.append((conf_matrix[i][j], i, j))

errors.sort(reverse=True)
most_errors = errors[:20]  # выведем 20 ошибок модели
labelnames = pd.read_csv('../data/label_names.csv')
for err in most_errors:
    name_of_most_errors = (labelnames['SignName'].iloc[err[1]], labelnames['SignName'].iloc[err[2]])
    print(name_of_most_errors)

('Speed limit (30km/h)', 'Speed limit (50km/h)')
('Speed limit (30km/h)', 'Speed limit (70km/h)')
('Speed limit (80km/h)', 'Speed limit (100km/h)')
('Speed limit (50km/h)', 'Speed limit (80km/h)')
('End of speed limit (80km/h)', 'Double curve')
('Speed limit (70km/h)', 'No vehicles')
('Speed limit (60km/h)', 'No passing')
('Speed limit (100km/h)', 'Speed limit (120km/h)')
('Speed limit (60km/h)', 'Speed limit (80km/h)')
('Speed limit (50km/h)', 'No entry')
('Speed limit (30km/h)', 'Speed limit (80km/h)')
('Slippery road', 'Beware of ice/snow')
('Speed limit (60km/h)', 'Keep right')
('No passing', 'No vehicles')
('Speed limit (100km/h)', 'Roundabout mandatory')
('Speed limit (80km/h)', 'End of speed limit (80km/h)')
('Speed limit (70km/h)', 'Traffic signals')
('Speed limit (60km/h)', 'Speed limit (100km/h)')
('Dangerous curve to the left', 'Road narrows on the right')
('Dangerous curve to the left', 'Double curve')


##### Модель чаще всего путает скорость, то есть плохо распознает цифры

# Medium

In [65]:
from sklearn.svm import SVC
from sklearn.naive_bayes import GaussianNB
from sklearn.neighbors import KNeighborsClassifier
from sklearn.metrics import f1_score

In [73]:
model = LogisticRegression(max_iter=200)
print(f"training of Logistic Regression in process...")
model.fit(dataframe_train, train['labels'])
print("training finished")
y_pred = model.predict(dataframe_test)
results = {'accuracy': accuracy_score(test['labels'], y_pred),
                     'precision': precision_score(test['labels'], y_pred, average='weighted', zero_division=0),
                     'recall': recall_score(test['labels'], y_pred, average='weighted'),
                     'f1': f1_score(test['labels'], y_pred, average='weighted', zero_division=0)}
print(f'Logistic Regression \n{results}')

training of Logistic Regression in process...
training finished
Logistic Regression 
{'accuracy': 0.8462391132224861, 'precision': 0.8511121516156515, 'recall': 0.8462391132224861, 'f1': 0.8450015707450316}


/Users/elizabeth/anaconda3/lib/python3.11/site-packages/sklearn/linear_model/_logistic.py:460: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  n_iter_i = _check_optimize_result(


In [66]:
model = GaussianNB()
print(f"training of NB in process...")
model.fit(dataframe_train, train['labels'])
print("training finished")
y_pred = model.predict(dataframe_test)
results = {'accuracy': accuracy_score(test['labels'], y_pred),
                     'precision': precision_score(test['labels'], y_pred, average='weighted', zero_division=0),
                     'recall': recall_score(test['labels'], y_pred, average='weighted'),
                     'f1': f1_score(test['labels'], y_pred, average='weighted', zero_division=0)}
print(f'NB \n{results}')

training of NB in process...
training finished
NB 
{'accuracy': 0.11971496437054632, 'precision': 0.3369179056192524, 'recall': 0.11971496437054632, 'f1': 0.13146324342826704}


In [74]:
model = SVC(kernel='rbf', max_iter=200)
print(f"training of SVM_rbf in process...")
model.fit(dataframe_train, train['labels'])
print("training finished")
y_pred = model.predict(dataframe_test)
results = {'accuracy': accuracy_score(test['labels'], y_pred),
                     'precision': precision_score(test['labels'], y_pred, average='weighted', zero_division=0),
                     'recall': recall_score(test['labels'], y_pred, average='weighted'),
                     'f1': f1_score(test['labels'], y_pred, average='weighted', zero_division=0)}
print(f'SVM_rbf \n{results}')

training of SVM_rbf in process...


/Users/elizabeth/anaconda3/lib/python3.11/site-packages/sklearn/svm/_base.py:297: ConvergenceWarning: Solver terminated early (max_iter=200).  Consider pre-processing your data with StandardScaler or MinMaxScaler.
  assert self.fit_status_ in (0, 1)


training finished
SVM_rbf 
{'accuracy': 0.5581947743467933, 'precision': 0.5873957719050017, 'recall': 0.5581947743467933, 'f1': 0.5443641621503109}


In [75]:
model = SVC(kernel='poly', max_iter=200)
print(f"training of SVM_poly in process...")
model.fit(dataframe_train, train['labels'])
print("training finished")
y_pred = model.predict(dataframe_test)
results = {'accuracy': accuracy_score(test['labels'], y_pred),
                     'precision': precision_score(test['labels'], y_pred, average='weighted', zero_division=0),
                     'recall': recall_score(test['labels'], y_pred, average='weighted'),
                     'f1': f1_score(test['labels'], y_pred, average='weighted', zero_division=0)}
print(f'SVM_poly \n{results}')

training of SVM_poly in process...


/Users/elizabeth/anaconda3/lib/python3.11/site-packages/sklearn/svm/_base.py:297: ConvergenceWarning: Solver terminated early (max_iter=200).  Consider pre-processing your data with StandardScaler or MinMaxScaler.
  assert self.fit_status_ in (0, 1)


training finished
SVM_poly 
{'accuracy': 0.26357878068091845, 'precision': 0.4748619164172877, 'recall': 0.26357878068091845, 'f1': 0.2745033824195049}


In [76]:
model = SVC(kernel='linear', max_iter=200)
print(f"training of SVM_linear in process...")
model.fit(dataframe_train, train['labels'])
print("training finished")
y_pred = model.predict(dataframe_test)
results = {'accuracy': accuracy_score(test['labels'], y_pred),
                     'precision': precision_score(test['labels'], y_pred, average='weighted', zero_division=0),
                     'recall': recall_score(test['labels'], y_pred, average='weighted'),
                     'f1': f1_score(test['labels'], y_pred, average='weighted', zero_division=0)}
print(f'SVM_linear \n{results}')

training of SVM_linear in process...


/Users/elizabeth/anaconda3/lib/python3.11/site-packages/sklearn/svm/_base.py:297: ConvergenceWarning: Solver terminated early (max_iter=200).  Consider pre-processing your data with StandardScaler or MinMaxScaler.
  assert self.fit_status_ in (0, 1)


training finished
SVM_linear 
{'accuracy': 0.7292161520190024, 'precision': 0.7461852845991985, 'recall': 0.7292161520190024, 'f1': 0.7299316466549948}


In [78]:
model = KNeighborsClassifier()
print(f"training of kNN in process...")
model.fit(dataframe_train, train['labels'])
print("training finished")
y_pred = model.predict(dataframe_test.values)
results = {'accuracy': accuracy_score(test['labels'], y_pred),
                     'precision': precision_score(test['labels'], y_pred, average='weighted', zero_division=0),
                     'recall': recall_score(test['labels'], y_pred, average='weighted'),
                     'f1': f1_score(test['labels'], y_pred, average='weighted', zero_division=0)}
print(f'kNN \n{results}')

training of kNN in process...
training finished
kNN 
{'accuracy': 0.3847189231987332, 'precision': 0.423392251068593, 'recall': 0.3847189231987332, 'f1': 0.39095958617170934}


# Итоги
1) Самые лучшие результаты у моделей svm_linear (accuracy ~ 0.72) и у логической регрессии (accuracy ~ 0.84)
2) svm_poly и svm_rbf обучались дольше всего и имеют не очень хорошие результаты
3) NB хорошо работает, если все признаки являются независимыми, а здесь это не выполнено, поэтому точность мала (accuracy ~ 0.11)
4) kNN зависит от размерности данных